src/componententity.cpp

```diff
@@ -246,34 +333,83 @@ size_t ComponentEntity::componentCount() const
     return mPimpl->mComponents.size();
 }
 
-bool ComponentEntity::containsComponent(const std::string &name) const
+bool ComponentEntity::containsComponent(const std::string &name, bool searchEncapsulated) const
 {
-    auto result = mPimpl->findComponent(name);
-    return result != mPimpl->mComponents.end();
+    bool result = false;
+    if (containsComponentInThis(name)) {
+        result = true;
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            result = getComponent(i)->containsComponent(name);
+            if (result) {
+                break;
+            }
+        }
+    }
+    return result;
 }
 
-ComponentPtr ComponentEntity::getComponent(size_t index)
+bool ComponentEntity::containsComponent(const ComponentPtr &component, bool searchEncapsulated) const
 {
-    return mPimpl->mComponents.at(index);
+    bool result = false;
+    if (containsComponentInThis(component)) {
+        result = true;
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            result = getComponent(i)->containsComponent(component);
```

Comment:
Don't you need to pass searchEncapsulated to the recursive call (here and elsewhere)?